### PR TITLE
[INTEL MKL] Relu op MKL-DNN 1.x integration

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**

### CODE REVIEW

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR!

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 1]
```diff
@@ -130,31 +154,47 @@ class MklEltwiseFwdPrimitive : public MklPrimitive {
...
-        new memory::primitive_desc(*context_.src_md, cpu_engine_));
 
-    // create a eltwise
-    context_.fwd_desc.reset(new mkldnn::eltwise_forward::desc(
+    context_.src_mpd.reset(
+#ifdef ENABLE_MKLDNN_V1
+        new MEMORY_PRIMITIVE_DESC(*context_.src_md));
+#else
+        new MEMORY_PRIMITIVE_DESC(*context_.src_md, cpu_engine_));
+#endif
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_relu_op.cc]: Done. Thanks!

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 2]
```diff
@@ -194,6 +247,29 @@ class MklEltwiseFwdPrimitiveFactory : public MklPrimitiveFactory<T> {
   MklEltwiseFwdPrimitiveFactory() {}
   ~MklEltwiseFwdPrimitiveFactory() {}
 
+#ifdef ENABLE_MKLDNN_V1
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_relu_op.cc]: Done. Thanks!

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 3]
```diff
@@ -170,6 +210,17 @@ class MklEltwiseFwdPrimitiveFactory : public MklPrimitiveFactory<T> {
...
+    eltwise_forward = static_cast<MklEltwiseFwdPrimitive<T>*>(
+        MklEltwiseFwdPrimitiveFactory<T>::GetInstance().GetEltwiseFwd(
+            fwdParams));
+    if (eltwise_forward == nullptr) {
+      eltwise_forward = new MklEltwiseFwdPrimitive<T>(fwdParams);
+      MklEltwiseFwdPrimitiveFactory<T>::GetInstance().SetEltwiseFwd(
+          fwdParams, eltwise_forward);
+    }
+#else
     auto src_fmt =
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_relu_op.cc]: Done. Thanks!

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 4]
```diff
@@ -333,42 +431,64 @@ class MklEltwiseBwdPrimitive : public MklPrimitive {
...
 
   // Eltwise backward primitive setup
   void Setup(const MklEltwiseBwdParams<T>& bwdParams) {
-    // create memory descriptors for eltwise data w/ no specified format
+    // Create memory descriptors for eltwise data w/ no specified format
     context_.src_md.reset(new memory::desc(bwdParams.common_md.data));
     context_.diff_dst_md.reset(new memory::desc(bwdParams.common_md.data));
 
+#ifdef ENABLE_MKLDNN_V1
+    context_.src_mpd.reset(new MEMORY_PRIMITIVE_DESC(*context_.src_md));
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_relu_op.cc]: Done. Thanks!

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 5]
```diff
@@ -130,31 +154,47 @@ class MklEltwiseFwdPrimitive : public MklPrimitive {
...
         fwdParams.alpha, fwdParams.beta));
-    context_.fwd_pd.reset(new mkldnn::eltwise_forward::primitive_desc(
-        *context_.fwd_desc, cpu_engine_));
-
-    // create memory primitive based on dummy data
+    context_.fwd_pd.reset(new EltwiseFwdPd(*context_.fwd_desc, cpu_engine_));
+    auto fwd_pd = context_.fwd_pd.get();
+
+#ifdef ENABLE_MKLDNN_V1
+    // Create memory primitive based on dummy data
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_relu_op.cc]: Although the code looks similar, it is actually different. So we don't get much benefit from code reuse if we refactor it into a function. Memory (for inputs/outputs) is different for every op as well as for fwd/bwd. Also, context_ is different for fwd/bwd.

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 6]
```diff
@@ -412,6 +546,29 @@ class MklEltwiseBwdPrimitiveFactory : public MklPrimitiveFactory<T> {
   }
 
  private:
+#ifdef ENABLE_MKLDNN_V1
+  static string CreateKey(const MklEltwiseBwdParams<T>& bwdParams) {
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_relu_op.cc]: In fact, MKL-DNN now supports primitive caching internally so it does not need to be handled at the user level. And the plan is to remove the primitive caching all together soon. So leaving it as is for now.

**Comment by mahmoud-abuzaina** [comment]
@penpornk Thank you for reviewing the PR. I have addressed your comments. Please have a look.

**Comment by Jianhui-Li** [comment]
@penpornk Appreciate if you can prioritize the review/merge-in this PR and a couple of others related to DNNL1.x upgrade. Thanks.

File: tensorflow/core/kernels/mkl_relu_op.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/mkl_relu_op.cc]: Got it. Thank you for the info!

File: tensorflow/core/kernels/mkl_relu_op.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/mkl_relu_op.cc]: Thank you for the clarification!

**Review by penpornk — APPROVED** [review]
Thank you for the quick fixes! I'll try to get to other PRs later this afternoon/tonight.

**Pull request merged**